src/group/libp2p_group_gossip_server.erl

```diff
@@ -1,105 +1,214 @@
-
 -define(DEFAULT_PEERBOOK_CONNECTIONS, 5).
+-define(DEFAULT_MAX_INBOUND_CONNECTIONS, 10).
 -define(DEFAULT_DROP_TIMEOUT, 5 * 60 * 1000).
 -define(GROUP_ID, "gossip").
+-define(GROUP_PATH, "gossip/1.0.0").
 
-%% gen_server
+%% API
 %%
 
 start_link(Sup, TID) ->
     gen_server:start_link(?MODULE, [Sup, TID], []).
 
+%% libp2p_gossip_stream
+%%
+
+handle_data(Pid, Key, Bin) ->
+    gen_server:call(Pid, {handle_data, Key, Bin}).
+
+accept_stream(Pid, SessionPid, StreamPid) ->
+    gen_server:call(Pid, {accept_stream, SessionPid, StreamPid}).
+
+
+%% gen_server
+%%
+
 init([Sup, TID]) ->
     erlang:process_flag(trap_exit, true),
-    %% TODO: Remove the assumption that this is _the_ group agent for
-    %% the swarm. Perhaps by moving the creation into the swarm server
-    %% instead of directly in the swarm_sup?
-    libp2p_swarm_sup:register_group_agent(TID),
+    libp2p_swarm_sup:register_gossip_group(TID),
     Opts = libp2p_swarm:opts(TID),
-    PeerBookCount = libp2p_group_gossip:get_opt(Opts, peerbook_connections, ?DEFAULT_PEERBOOK_CONNECTIONS),
-    DropTimeOut = libp2p_group_gossip:get_opt(Opts, drop_timeout, ?DEFAULT_DROP_TIMEOUT),
-    ClientSpec = libp2p_group_gossip:get_opt(Opts, stream_client, undefined),
-    SeedNodes = libp2p_group_gossip:get_opt(Opts, seed_nodes, []),
+    PeerBookCount = get_opt(Opts, peerbook_connections, ?DEFAULT_PEERBOOK_CONNECTIONS),
+    InboundCount = get_opt(Opts, inbound_connections, ?DEFAULT_MAX_INBOUND_CONNECTIONS),
+    DropTimeOut = get_opt(Opts, drop_timeout, ?DEFAULT_DROP_TIMEOUT),
+    SeedNodes = get_opt(Opts, seed_nodes, []),
     self() ! {start_workers, PeerBookCount, length(SeedNodes)},
-    {ok, update_metadata(#state{sup=Sup, tid=TID, client_spec=ClientSpec,
-                                seed_nodes=SeedNodes, peerbook_connections=PeerBookCount,
-                                drop_timeout=DropTimeOut, drop_timer=schedule_drop_timer(DropTimeOut)})}.
+    {ok, update_metadata(#state{sup=Sup, tid=TID,
+                                seed_nodes=SeedNodes,
+                                max_inbound_connections=InboundCount,
+                                peerbook_connections=PeerBookCount,
+                                drop_timeout=DropTimeOut,
+                                drop_timer=schedule_drop_timer(DropTimeOut)})}.
+
+handle_call({accept_stream, _Session, _StreamPid}, _From, State=#state{workers=[]}) ->
+    {reply, {error, not_ready}, State};
+handle_call({accept_stream, Session, StreamPid}, From, State=#state{}) ->
+    libp2p_session:identify(Session, self(), {From, StreamPid}),
+    {noreply, State};
+handle_call({handle_data, Key, Msg}, _From, State=#state{}) ->
+    %% Incoming message from a gossip stream for a given key
+    case maps:find(Key, State#state.handlers) of
+        error -> {reply, ok, State};
+        {ok, {M, S}} -> case (catch M:handle_gossip_data(Msg, S)) of
+                               {'EXIT', Err} -> {reply, {error, Err}, State};
```

Comment:
I think the catches should have some explanation as to why they're needed.

Reply:
I get it now though, the modules are installed into this handler like a gen_event and you don't want one bad gossip handler breaking the whole gossip server.

Reply:
I'll add a comment to that effect